neo/Ledger/Blockchain.cs

```diff
@@ -409,11 +439,113 @@ private RelayResultReason OnNewTransaction(Transaction transaction)
+                        {
+                            system.LocalNode.Tell(new LocalNode.SendDirectly { Inventory = state_root_verifying });
+                        }
+                    }
+                    else
+                    {
+                        local_state.Flag = StateRootVerifyFlag.Invalid;
+                    }
+                    snapshot.Commit();
+                    UpdateCurrentSnapshot();
+                    if (local_state.Flag == StateRootVerifyFlag.Invalid) break;
+                }
+            }
+            return RelayResultReason.Succeed;
+        }
+
+        private void OnStateRoots(StateRoot[] sts)
+        {
+            foreach (var state_root in sts)
+            {
+                var result = OnNewStateRoot(state_root);
+                if (result != RelayResultReason.Succeed)
+                {
+                    break;
+                }
+            }
+            system.TaskManager.Tell(new TaskManager.StateRootTaskCompleted(), Sender);
+        }
+
+        private void OnImportRoots(IEnumerable<StateRoot> roots)
+        {
+            foreach (StateRoot root in roots)
+            {
+                if (root.Index < Math.Max(StateHeight, StateRootEnableIndex)) continue;
+                if (root.Index != Math.Max(StateHeight + 1, StateRootEnableIndex))
+                    throw new InvalidOperationException();
+                using (Snapshot snapshot = GetSnapshot())
+                {
+                    var local_state = snapshot.StateRoots.GetAndChange(root.Index);
+                    if (local_state.Flag == StateRootVerifyFlag.Invalid) break;
+                    if (local_state.StateRoot.Root == root.Root && local_state.StateRoot.PreHash == root.PreHash)
+                    {
+                        snapshot.StateRootHashIndex.GetAndChange().Index = root.Index;
+                        snapshot.StateRootHashIndex.GetAndChange().Hash = root.Hash;
+                        local_state.StateRoot = root;
+                        local_state.Flag = StateRootVerifyFlag.Verified;
+                    }
+                    else
+                    {
+                        local_state.Flag = StateRootVerifyFlag.Invalid;
+                    }
+                    snapshot.Commit();
+                    UpdateCurrentSnapshot();
+                    if (local_state.Flag == StateRootVerifyFlag.Invalid) break;
+                }
+            }
+            Sender.Tell(new ImportCompleted());
+        }
+
         private void OnPersistCompleted(Block block)
         {
             block_cache.Remove(block.Hash);
             MemPool.UpdatePoolForBlockPersisted(block, currentSnapshot);
             Context.System.EventStream.Publish(new PersistCompleted { Block = block });
+            var index = Math.Max(StateHeight + 1, StateRootEnableIndex);
+            if (GetStateRoot(index)?.Flag == StateRootVerifyFlag.Unverified && state_root_cache.ContainsKey(index))
```

Comment:
```suggestion
            if (GetStateRoot(index)?.Flag == StateRootVerifyFlag.Unverified && state_root_cache.TryRemove(index, out StateRoot state_root))
```

Reply:
Seems `Dictionary` doesn't have this method.